### PR TITLE
Namespaced Operator

### LINKED ISSUE #871: Dynamic Engine version support at Seldon operator
Based on the understanding of the current code (0.4.0), the engine is an environment variable of Seldon operator. Whenever a SeldonDeployment is deployed, the engine defined at the ENV is used for the deployment pod. Every deployment that uses this operator will have the same engine.  Is there a way that the SeldonDeployment can specify an Engine version? I mean different SeldonDeployment can use different Engine version while still share the Operator. For example,  let's say the default environment variable configured engine version is 0.4.0 and the new version I plan to test is 0.4.1. If I could specify an Engine version at my SeldonDeployment, my specific deployment pod uses 0.4.1 and other SeldonDeployments still use the configured 0.4.0. The benefit of using dynamic Engine version can facilitate engine customization testing/development at customer deployment without affecting the operation.

### LINKED ISSUE #1139: Seldon Core Operator Restricted to Single Namespace
This task encompasses the functionality to deploy the Seldon Core Operator restricted to a single namespace. Our operator is built using Kubebuilder which by default was using Cluster-wide roles, however since we upgraded to Kubebuilder 2.0 we are now able to use normal roles, which would enable us to have more granular restrictions. This is also something that has been requested previously.

Related issues include #670, #890

### PR DESCRIPTION
**Pull request by cliveseldon**
Fixes #1139 
Fixes #871 

 * Allow namespaced Operator
 * Allow Operator that manages only resources with `controler-id` matching label

Includes:

 * Updated install docs
 * Example notebook showing clusterwide, namespaced and labelled operators all running together.

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **cliveseldon**

If they are not already assigned, you can assign the PR to them by writing `/assign @cliveseldon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by cliveseldon** [comment]
Need to ensure webhook config is scoped: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules

**Comment by phsiao** [comment]
How to prevent the cluster-wide operator from competing with namespaced operator?

**Comment by cliveseldon** [comment]
> How to prevent the cluster-wide operator from competing with namespaced operator?

Good question. Is that good practice to have a cluster wide and namespaced operators in same cluster?

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | d7d0b386806c959d595eea02451ea6f2de852f84 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | d7d0b386806c959d595eea02451ea6f2de852f84 | [link]() | `/test lint`
pr-build | d193d0c0d68f5bfb2f4237704594f2745055ff09 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | d193d0c0d68f5bfb2f4237704594f2745055ff09 | [link]() | `/test this`
lint | d193d0c0d68f5bfb2f4237704594f2745055ff09 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | d193d0c0d68f5bfb2f4237704594f2745055ff09 | [link]() | `/test lint`
pr-build | 43c1de93328394e5483a6ee353e918be7090cc4d | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 43c1de93328394e5483a6ee353e918be7090cc4d | [link]() | `/test this`
lint | 43c1de93328394e5483a6ee353e918be7090cc4d | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 43c1de93328394e5483a6ee353e918be7090cc4d | [link]() | `/test lint`
pr-build | bab53b2e534450069fa410435b33e32a6cf7b053 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | bab53b2e534450069fa410435b33e32a6cf7b053 | [link]() | `/test this`
lint | bab53b2e534450069fa410435b33e32a6cf7b053 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: operator/controllers/seldondeployment_controller.go [hunk 1]
```diff
@@ -52,18 +52,21 @@ import (
 const (
 	ENV_DEFAULT_ENGINE_SERVER_PORT      = "ENGINE_SERVER_PORT"
 	ENV_DEFAULT_ENGINE_SERVER_GRPC_PORT = "ENGINE_SERVER_GRPC_PORT"
+	ENV_CONTROLLER_ID                   = "CONTROLLER_ID"
 
 	DEFAULT_ENGINE_CONTAINER_PORT = 8000
 	DEFAULT_ENGINE_GRPC_PORT      = 5001
 
 	AMBASSADOR_ANNOTATION = "getambassador.io/config"
+	LABEL_CONTROLLER_ID   = "controller-id"
```
Comment (phsiao) [inline operator/controllers/seldondeployment_controller.go]: I would suggest adding some prefix to "controller-id" for label key, such as "seldon.io/controller-id" to distinguish it from possible label naming collision.

**Comment by seldondev** [comment]
Fri Dec  6 16:34:10 UTC 2019
The logs for [pr-build] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=14

**Comment by seldondev** [comment]
Fri Dec  6 16:34:16 UTC 2019
The logs for [lint] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=15

File: operator/controllers/seldondeployment_controller.go [hunk 2]
```diff
@@ -1218,6 +1216,29 @@ func (r *SeldonDeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
...
 
+	// Check if we are not namespaced and should ignore this as its in a namespace managed by another operator
+	if r.Namespace == "" {
+		ns := &corev1.Namespace{}
+		err := r.Get(ctx, types.NamespacedName{Name: instance.Namespace}, ns)
+		if err != nil {
+			log.Error(err, "unable to fetch SeldonDeployment namespace", "namespace", instance.Namespace)
+			return ctrl.Result{}, err
+		} else {
+			if ns.Labels[LABEL_CONTROLLER_ID] != "" {
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/controllers/seldondeployment_controller.go]: The check is to stop a clusterwide Operator from also executing on the create of a resource in a namespace in which there is also a seldon controller.

[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/controllers/seldondeployment_controller.go]: The webhooks don't get called due to the check on the namespace but I seem to have to add a check for Reconcile as that still gets called by the k9s-api.

[the remark this answers is not in the document]
Comment (phsiao) [reply operator/controllers/seldondeployment_controller.go]: You are right, I forgot about that path.

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | d602c63e4de6dfb1027e5f8779152efd13624351 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | d602c63e4de6dfb1027e5f8779152efd13624351 | [link]() | `/test this`
lint | d602c63e4de6dfb1027e5f8779152efd13624351 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Sat Dec  7 09:57:04 UTC 2019
The logs for [pr-build] [18] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/18.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=18

**Comment by seldondev** [comment]
Sat Dec  7 09:57:08 UTC 2019
The logs for [lint] [19] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/19.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=19

**Comment by seldondev** [comment]
Sat Dec  7 10:39:39 UTC 2019
The logs for [pr-build] [20] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/20.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=20

**Comment by seldondev** [comment]
Sat Dec  7 10:39:46 UTC 2019
The logs for [lint] [21] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/21.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=21

**Comment by seldondev** [comment]
Mon Dec  9 11:56:14 UTC 2019
The logs for [pr-build] [22] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/22.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=22

**Comment by seldondev** [comment]
Mon Dec  9 11:56:18 UTC 2019
The logs for [lint] [23] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/23.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=23

File: operator/config/rbac/kustomization.yaml [hunk 3]
```diff
@@ -6,9 +6,9 @@ resources:
 # Comment the following 3 lines if you want to disable
 # the auth proxy (https://github.com/brancz/kube-rbac-proxy)
 # which protects your /metrics endpoint.
-- auth_proxy_service.yaml
-- auth_proxy_role.yaml
-- auth_proxy_role_binding.yaml
+#- auth_proxy_service.yaml
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/config/rbac/kustomization.yaml]: yes. Wanted to simplify the install by removing the proxy auth.

**Comment by seldondev** [comment]
Mon Dec  9 13:13:43 UTC 2019
The logs for [pr-build] [24] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/24.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=24

**Comment by seldondev** [comment]
Mon Dec  9 13:14:10 UTC 2019
The logs for [integration] [25] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/25.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=25

**Comment by seldondev** [comment]
Mon Dec  9 13:14:11 UTC 2019
The logs for [lint] [26] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/26.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=26

**Comment by seldondev** [comment]
Mon Dec  9 15:12:38 UTC 2019
The logs for [pr-build] [27] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/27.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=27

**Comment by seldondev** [comment]
Mon Dec  9 15:12:47 UTC 2019
The logs for [lint] [28] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/28.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=28

**Comment by seldondev** [comment]
Mon Dec  9 20:45:33 UTC 2019
The logs for [pr-build] [29] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/29.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=29

**Comment by seldondev** [comment]
Mon Dec  9 20:45:43 UTC 2019
The logs for [lint] [30] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1142/30.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1142 --build=30

**Pull request merged**